crates/ark/src/interface.rs

```diff
@@ -1735,7 +1735,9 @@ impl RMain {
 
     fn send_lsp_notification(&self, event: KernelNotification) {
         if let Some(ref tx) = self.lsp_events_tx {
-            tx.send(Event::Kernel(event)).unwrap();
+            if let Err(err) = tx.send(Event::Kernel(event)) {
```

Comment:
This is the main change that fixes the crash right?

To avoid multiple logged errors, we should also set `self.lsp_events_tx` to `None` when it's dropped. A bit tricky because this requires access to `RMain`.

Ideally we'd change this to a a broadcast channel so that multiple clients could be notified. And a broadcast would always succeed no matter the number of connected clients (0, 1, or more).

Perhaps we should do that now even if we don't support multiple clients yet?

Reply:
> Ideally we'd change this to a a broadcast channel so that multiple clients could be notified.

That's an interesting idea! I started down this road, but it began metastasizing into a much bigger change than I wanted. In particular the LSP events use a MPSC model where there are many copies of the `Sender` passed around to the different spots in the code where we need to send notifications. The `Bus` broadcast model is single producer, multi consumer -- you can't `clone()` a `Bus` or generate additional senders. We _could_ pass around a Arc/Mutex containing a Bus instance, or use a MPSC => SPMC design to fan in/fan out the messages, but that's all adding more complexity than I think this change warrants.

Reply:
> Ideally we'd change this to a a broadcast channel so that multiple clients could be notified.

Actually it's not possible for multiple clients to be connected at the same time, the LSP protocol does not support it:

> The protocol currently assumes that one server serves one tool. There is currently no support in the protocol to share one server between different tools. Such a sharing would require additional protocol e.g. to lock a document to support concurrent editing.

So ideally we'd shut down the currently running LSP when we get a request to create a new one.

Reply:
Well we can't really shut down a server from the server.

> The current protocol specification defines that the lifecycle of a server is managed by the client (e.g. a tool like VS Code or Emacs). It is up to the client to decide when to start (process-wise) and when to shutdown a server.

So I guess we just let the zombie server run 🤷

Reply:
This `AUXILIARY_EVENT_TX` global is a bit tricky. It's a global so that LSP methods can log messages from free functions/macros. That's purely for ergonomics of development so we don't have to pass around a logger handle everywhere we want to log something.

And we can't make it thread local since the LSP runs on a thread pool.